File: LeetCode/0171-Excel Sheet Column Number/main.cpp
//
// Created by 薛智钧 on 2020/3/31.
//
#include <string>
#include <iostream>
#include <math.h>
using namespace std;

/*
 * Given a column title as appear in an Excel sheet, return its corresponding column number.

   For example:

    A -> 1
    B -> 2
    C -> 3
    ...
    Z -> 26
    AA -> 27
    AB -> 28
*/

class Solution {
public:
    // 从低位到高位
    int titleToNumber(string s) {
        if (s.empty()) return 0;
        int res = 0;
        auto iter = s.rbegin();
        for (int i = 0; i < s.size(); ++i, ++iter){
            int digit = *iter - 'A' + 1;
            res += (int)(digit * pow(26, i));
        }
        return res;
    }

    // 从高位到低位
    int titleToNumber_2(string s){
        int res = 0;
        for (char i : s)
            res = res * 26 + (i - 'A' + 1);
        return res;
    }
};

int main(){
    string s = "CU";
    Solution sol;
    cout << sol.titleToNumber_2(s);
}

File: LeetCode/0171-Excel Sheet Column Number/main.py
"""  
Given a column title as appear in an Excel sheet, return its corresponding column number.

For example:

    A -> 1
    B -> 2
    C -> 3
    ...
    Z -> 26
    AA -> 27
    AB -> 28 
    ...
Example 1:

Input: "A"
Output: 1
Example 2:

Input: "AB"
Output: 28
Example 3:

Input: "ZY"
Output: 701
"""

# ord(char) = char的ASCII码值
class Solution:
    def titleToNumber(self, s: str) -> int:
        res = 0
        for i in s:
            res = res * 26 + (ord(i) - 65 + 1) # ord('A') = 65
            
        return res